### PR TITLE
fix: trends curve fitting issue

### PR DESCRIPTION
**Pull request by Inglezos**
Fixed trends curve fitting issue using log(S) and linear function.
Also I changed min_size to 5.

## Related issues/bugs
This is related to #282.

## New features
Added linear function for trends fitting.

### CODE REVIEW

File: covsirphy/cleaning/term.py [hunk 1]
```diff
@@ -128,6 +128,21 @@ def negative_exp(x, a, b):
             float
         """
         return a * np.exp(-b * x)
+    
+    @staticmethod
+    def linear_fnc(x, a, b):
```
Comment (lisphilar) [inline covsirphy/cleaning/term.py:133]: Thank you!

File: covsirphy/phase/sr_change.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ class ChangeFinder(Term):
         When RMSLE score > max_rmsle, predicted values will be None
     """
 
-    def __init__(self, sr_df, min_size=7, max_rmsle=20.0):
+    def __init__(self, sr_df, min_size=5, max_rmsle=20.0):
```
Comment (lisphilar) [inline covsirphy/phase/sr_change.py:31]: I confirm this change is necessary.

File: covsirphy/phase/sr_change.py [hunk 3]
```diff
@@ -129,8 +129,17 @@ def _curve_fitting(self, phase, start_date, end_date):
         sr_df = sr_df.loc[(sr_df.index >= sta) & (sr_df.index <= end), :]
         trend = Trend(sr_df)
         df = trend.run()
+        if trend.rmsle() > trend.max_linear_fit_err:
```
Comment (lisphilar) [inline covsirphy/phase/sr_change.py:132]: Please consider to perform both of linear regression and curve fitting with exponential function. Then, select better function using RMSLE score.

```Python
# Linear function
trend_line = Trend(sr_df)
line_df = trend_line.run(func="linear")
# Negative exponential function
trend_exp = Trend(sr_df)
exp_df = trend_exp.run(func="negative_exppnential")
# Compare the trends with RMSLE score
df = line_df.copy() if trend_line.rmsle() < trend_exp.rmsle() else exp_df.copy()
```

File: covsirphy/phase/sr_change.py [hunk 4]
```diff
@@ -129,8 +129,17 @@ def _curve_fitting(self, phase, start_date, end_date):
...
+            # if the linear curve fitting error is significant
+            # then rerun the trend analysis using
+            # the negative exponential function for fitting
+            trend = Trend(sr_df)
+            trend.use_neg_exp_fnc = 1
+            df = trend.run()
+            trend.use_neg_exp_fnc = 0
         if trend.rmsle() > self.max_rmsle:
             df[f"{self.S}{self.P}"] = None
+        # print(trend.rmsle())
```
Comment (lisphilar) [inline covsirphy/phase/sr_change.py:142]: Please remove line 142.

File: covsirphy/phase/trend.py [hunk 5]
```diff
@@ -36,6 +36,9 @@ def __init__(self, sr_df):
             raise ValueError("The length of @sr_df must be over 2.")
         # Setting for analysis
         self.result_df = None
+        self.max_linear_fit_err = 0.5
```
Comment (lisphilar) [inline covsirphy/phase/trend.py:39]: Please create `Trend._run_linear()` and `Trend._run_negative_exponential()`. Then, add `func` argument to `Trend.run()`.

**Review by lisphilar — CHANGES_REQUESTED** [review]
Thank you for your pull request!!
I agree with your idea to use linear regression, but please make some changes.

1. Please consider to perform both of linear regression and curve fitting with exponential function. Then, select better function using RMSLE score.

2. Please `Trend._run_linear()` and `Trend._run_negative_exponential`.

3. Please update test codes in `tests/test_change_finder.py` in this repository to cover all codes.
CI tool will perform tests with `pytest` package.

File: covsirphy/phase/trend.py (hunk 5 above)
Comment (Inglezos) [reply covsirphy/phase/trend.py:39]: Why create these two functions? Isn't enough to add only in `__init__` of the `Trend` class a variable `self.fit_fnc` and set this to either `self.linear` or `self.negative_exp` depending on the `Trend.run(func)`?

Comment (lisphilar) [reply covsirphy/phase/trend.py]: Thank you, your solution (`.run(func)`) is better.

**Pull request merged**